controller/node_controller.go

```diff
@@ -1018,3 +1001,215 @@ func BackingImageDiskFileCleanup(node *longhorn.Node, bi *longhorn.BackingImage,
+			if !onDiskDataExist {
+				if err := nc.ds.DeleteOrphan(orphan.Name); err != nil && !apierrors.IsNotFound(err) {
+					return errors.Wrapf(err, "failed to delete orphan %v", orphan.Name)
+				}
+				continue
+			}
+		}
+
+		dataCleanableCondition := types.GetCondition(orphan.Status.Conditions, longhorn.OrphanConditionTypeDataCleanable)
+		if dataCleanableCondition.Status == longhorn.ConditionStatusUnknown {
+			continue
+		}
+
+		if autoDeletionEnabled || dataCleanableCondition.Status == longhorn.ConditionStatusFalse {
+			if err := nc.ds.DeleteOrphan(orphan.Name); err != nil && !apierrors.IsNotFound(err) {
+				return errors.Wrapf(err, "failed to delete orphan %v", orphan.Name)
+			}
+		}
+	}
+	return nil
+}
+
+func (nc *NodeController) isOnDiskDataExist(orphan *longhorn.Orphan, orphanedReplicaDirectoryNames map[string]string) (bool, error) {
+	replicaDirectoryName := orphan.Spec.Parameters[longhorn.OrphanDataName]
+	_, exist := orphanedReplicaDirectoryNames[replicaDirectoryName]
+
+	return exist, nil
+}
+
+func (nc *NodeController) createOrphans(node *longhorn.Node, diskName string, diskInfo *monitor.CollectedDiskInfo) error {
+	for dirName := range diskInfo.OrphanedReplicaDirectoryNames {
+		if err := nc.createOrphan(node, diskName, dirName, diskInfo); err != nil && !apierrors.IsAlreadyExists(err) {
+			return errors.Wrapf(err, "unable to create orphan for orphaned replica directory %v in disk %v on node %v",
+				dirName, node.Spec.Disks[diskName].Path, node.Name)
+		}
+	}
+
+	return nil
+}
+
+func (nc *NodeController) createOrphan(node *longhorn.Node, diskName, replicaDirectoryName string, diskInfo *monitor.CollectedDiskInfo) error {
+	name := types.GetOrphanChecksumNameForOrphanedDirectory(node.Name, diskName, diskInfo.DiskUUID, replicaDirectoryName)
+
+	_, err := nc.ds.GetOrphanRO(name)
+	if err == nil || (err != nil && !apierrors.IsNotFound(err)) {
+		return err
+	}
+
+	orphan := &longhorn.Orphan{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: longhorn.OrphanSpec{
+			NodeID: node.Name,
+			Type:   longhorn.OrphanTypeReplica,
+			Parameters: map[string]string{
+				longhorn.OrphanDataName: replicaDirectoryName,
+				longhorn.OrphanDiskName: diskName,
+				longhorn.OrphanDiskUUID: node.Status.DiskStatus[diskName].DiskUUID,
+				longhorn.OrphanDiskPath: node.Spec.Disks[diskName].Path,
+			},
+		},
+	}
+
+	_, err = nc.ds.CreateOrphan(orphan)
+
+	return err
+}
+
+func (nc *NodeController) syncWithMonitor(mon monitor.Monitor, node *longhorn.Node) (map[string]*monitor.CollectedDiskInfo, error) {
+	v, err := mon.GetCollectedData()
+	if err != nil {
+		return nil, err
+	}
+
+	collectedDiskInfo := v.(map[string]*monitor.CollectedDiskInfo)
+	if !isDiskMatched(node, collectedDiskInfo) {
+		return nil, errors.New("mismatching disks in node resource object and monitor collected data")
+	}
+	return collectedDiskInfo, nil
+}
+
+func isDiskMatched(node *longhorn.Node, collectedDiskInfo map[string]*monitor.CollectedDiskInfo) bool {
```

Comment:
Print out the mismatching reason.

BTW, the mismatching should be caused by the monitor syncing up the status before the disk spec update, right? And the mismatching should disappear in the next sync.

Reply:
> And the mismatching should disappear in the next sync.

yeah